amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -446,41 +455,128 @@ void runProcessLoop() {
+        final Set<StreamIdentifier> streamIdentifiers = leases.stream()
+                .map(lease -> StreamIdentifier.multiStreamInstance(lease.streamIdentifier()))
+                .collect(Collectors.toSet());
+        for (StreamIdentifier streamIdentifier : streamIdentifiers) {
+            if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
+                currentStreamConfigMap.put(streamIdentifier, getDefaultStreamConfig(streamIdentifier));
+            }
+        }
+    }
+
+    private List<MultiStreamLease> fetchMultiStreamLeases()
             throws DependencyException, ProvisionedThroughputException, InvalidStateException {
-        if (!leasesSyncedOnAppInit && isMultiStreamMode) {
-            final Set<StreamIdentifier> streamIdentifiers = leaseCoordinator.leaseRefresher().listLeases().stream()
-                    .map(lease -> StreamIdentifier.multiStreamInstance(((MultiStreamLease) lease).streamIdentifier()))
-                    .collect(Collectors.toSet());
-            for (StreamIdentifier streamIdentifier : streamIdentifiers) {
-                if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
-                    currentStreamConfigMap.put(streamIdentifier, getDefaultStreamConfig(streamIdentifier));
+        return (List<MultiStreamLease>) ((List) leaseCoordinator.leaseRefresher().listLeases());
+    }
+
+    private void removeActiveStreamsFromStaleStreamsList(Set<StreamIdentifier> streamIdentifiers) {
+        for(StreamIdentifier streamIdentifier : streamIdentifiers) {
+            staleStreamDeletionMap.remove(streamIdentifier);
+        }
+    }
+
+    private Set<StreamIdentifier> deleteMultiStreamLeases(Set<StreamIdentifier> streamIdentifiers)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final Set<StreamIdentifier> streamsSynced = new HashSet<>();
+        List<MultiStreamLease> leases = null;
+        Map<String, List<MultiStreamLease>> streamIdToShardsMap = null;
+        for(StreamIdentifier streamIdentifier : streamIdentifiers) {
+            if (leases == null) {
+                // Lazy Load once and use many times for this iteration.
+                leases = fetchMultiStreamLeases();
+            }
+            if (streamIdToShardsMap == null) {
+                // Lazy load once and use many times for this iteration.
+                streamIdToShardsMap = leases.stream().collect(Collectors
+                        .groupingBy(MultiStreamLease::streamIdentifier,
+                                Collectors.toCollection(ArrayList::new)));
+            }
+            log.warn("Found old/deleted stream: " + streamIdentifier + ". Deleting leases of this stream.");
+            // Deleting leases will cause the workers to shutdown the record processors for these shards.
+            if (deleteMultiStreamLeases(streamIdToShardsMap.get(streamIdentifier.serialize()))) {
```

Comment:
Is that the assumption that all the workers of the leases would have stopped processing ? What happens if worker1 deletes the lease of the stream which has ER disabled and another worker which is working on the leases and trying to checkpoint

Reply:
Deleting a lease, will cause the current worker holding the lease to give up the lease and shutdown the record processor. The worker trying to checkpoint will have the update operation failed, exception will be propagated to shard record processor and then to the respective task (processTask or shutdown task) and suppressed after logging.